kyaml/yaml/fns.go

```diff
@@ -63,17 +63,48 @@ type ElementSetter struct {
 	Value string `yaml:"value,omitempty"`
 }
 
+func (e ElementSetter) Filter(rn *RNode) (*RNode, error) {
+	return rn.Pipe(ElementSetterList{
+		Element: e.Element,
+		Keys:    []string{e.Key},
+		Values:  []string{e.Value},
+	})
+}
+
+// ElementSetterList sets the value for an Element in an associative list.
+// It behaves identically to ElementSetter, except that it uses multiple
+// key-value pairs (in the form of a list of keys and a corresponding list
+// of values) in order to find a matching element, whereas ElementSetter
+// uses only one key-value pair.
+type ElementSetterList struct {
+	Kind string `yaml:"kind,omitempty"`
+
+	// Element is the new value to set -- remove the existing element if nil
+	Element *Node
+
+	// Key is a list of fields on the elements. It is used to find matching elements to
+	// update / delete
+	Keys []string
```

Comment:
Is it necessary to put keys and values into different arrays? It looks like the `Keys[i]` is always paired with `Values[i]`. If you put them into different arrays, you may need to check do the lengths equal, otherwise line 132 may run out of range if length of Values is less than the length of Keys. An alternative is putting them into a 2d array.

Just a suggestion because using 2 arrays for a list of paired value makes thing more complex and you have to write codes to check them.

Reply:
I agree with what you're saying but I'm a little concerned that it will add significant complexity to associative_sequence.go. That file in setAssociativeSequenceElements currently has an array of values that enumerate all the values in the yaml corresponding to the key, e.g. [8080, 80]. #3159 will have to turn that into a 2D array so that each element of the values array has the value for each key, e.g. [ [8080, UDP], [8080, TCP] ]. 

My concern is if ElementSetterList decides to take a single 2D array instead of keys and values, then the two options I can think of for associative_sequence.go are, it will either need a 3D array, where each element is a 2D array containing all the key-value pairs. Or it will need to generate a 2D array to pass to ElementSetterList every time. I feel that adding logic to check that ElementSetterList keys/values are the same length will be less complex, what do you think?

Reply:
Checking length is OK for me and you should have test case to cover it.